### ISSUE TITLE
Implement old password authentication (pre-4.1 MySQL Server)

### ISSUE THREAD

**Issue by amacado** [mysql-net/MySqlConnector#856]
I need to support MySQL Server prior 4.1 (in this case 4.0) which throws me the error `System.NotSupportedException: Authentication method '' is not supported`.

```
System.NotSupportedException: Authentication method '' is not supported.
   at MySqlConnector.Core.ServerSession.ConnectAsync(ConnectionSettings cs, Int32 startTickCount, ILoadBalancer loadBalancer, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ServerSession.cs:line 399
   at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ConnectionPool.cs:line 110
   at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ConnectionPool.cs:line 140
   at MySqlConnector.MySqlConnection.CreateSessionAsync(ConnectionPool pool, Int32 startTickCount, Nullable`1 ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlConnection.cs:line 819
   at MySqlConnector.MySqlConnection.OpenAsync(Nullable`1 ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlConnection.cs:line 399
   at MySqlConnector.MySqlConnection.Open() in /_/src/MySqlConnector/MySqlConnection.cs:line 365
```

This problem is beeing referenced in #139:
> This would also be a good time to implement Old Password Authentication.
> This was not implemented, as it's only necessary for pre-4.1 MySQL Servers, which doesn't seem worth supporting.
_Originally posted by @bgrainger in https://github.com/mysql-net/MySqlConnector/issues/139#issuecomment-262624416_

Is there any update on this topic? What would be the required steps?

**Comment by bgrainger** [comment]
> What would be the required steps?

There would have to be code added to detect the absence of `ProtocolCapabilities.Protocol41` (and `PluginAuth`), then flag that as `mysql_old_password`, then support for `mysql_old_password` would have to be implemented.

The algorithm doesn't appear to be documented here, so correct documentation would have to be located: https://dev.mysql.com/doc/internals/en/old-password-authentication.html#packet-Authentication::Old

**Comment by amacado** [comment]
Thanks for your response @bgrainger, I might (have to) put my hands on it, but at first I will try my best to update the MySQL server. I'm responding back here if I will take the try to implement it, ok?

**Comment by bgrainger** [comment]
Upgrading MySQL Server to a supported version sounds like the best approach. According to https://en.wikipedia.org/wiki/MySQL#Release_history, MySQL 5.1 has been out of support since 2013, and 4.0 is so old it doesn't even appear on the chart.

Additionally, the reason this authentication scheme isn't supported is that it's insecure and can be hacked, so continuing to use it seems dangerous.

**Comment by bgrainger** [comment]
Actually, MySqlConnector assumes that `CLIENT_PROTOCOL_41` is supported, so it probably can't communicate at all with a MySQL 4.0 server.

**Comment by bgrainger** [comment]
MySQL Server 4.0 isn't supported because 1) old password authentication isn't implemented; 2) the 4.1 protocol (`CLIENT_PROTOCOL_41`) is assumed to always be on.

Fixing (1) itself might not be too hard, but implementing (2) would require changes in many places (EOF packets, column definitions, etc), and the documentation for the old protocol may be incomplete.

Since that server version is very old, insecure, and unsupported (and probably has a very small install base?), there's little value in writing a lot of extra code to try to be compatible with it; I would recommend updating the server to 5.7.x or 8.0.x.
